sensors/ev3_uart_sensor_defs.c

```diff
@@ -86,25 +86,67 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
 			},
 			[3] = {
 				/**
-				 * @description: Raw Reflected - sets LED color to red
-				 * @value0: ??? (0 to 1020???)
-				 * @value1: ??? (0 to 1020???)
+				 * .. [#lego-ev3-color-mode3-values] This mode
+				 *    provides direct access to the measurements
+				 *    done by the ADC inside the sensor. It
+				 *    works by continuously flashing the red LED
+				 *    on and off and measuring the current
+				 *    flowing through a bypass capacitor. It is
+				 *    not possible to measure true ambient light
+				 *    this way, but the LED off measurement
+				 *    works as a reference for the LED on
+				 *    measurement. Higher photodiode
+				 *    illumination will cause lower ADC
+				 *    measurements., so to obtain reflected
+				 *    light intensity similar to
+				 *    ``COL-REFLECT``, it is necessary to
+				 *    do ``reflect = value1 - value0``.
+				 *
+				 * @description: Raw Reflected Light - sets LED color to red
+				 * @value0: ADC reading for LED on (reflection) (0 to 1023)
+				 * @value0_footnote: [#lego-ev3-color-mode3-values]_
+				 * @value1: ADC reading for LED off (reference) (0 to 1023)
+				 * @value1_footnote: [#lego-ev3-color-mode3-values]_
 				 */
 				.name		= "REF-RAW",
 				.data_sets	= 2,
 				.data_type	= LEGO_SENSOR_DATA_S16,
+				.units		= "ADC",
 			},
 			[4] = {
 				/**
+				 * .. [#lego-ev3-color-mode4-value012] These
+				 *    values represent the difference between
+				 *    the raw sensor ADC measurements when the
+				 *    LEDs are off and when they are on. They
+				 *    work similar to the ``COL-REFLECT`` values
+				 *    but they are not precalibrated.
+				 *
+				 * .. [#lego-ev3-color-mode4-value3] This value
+				 *    is similar to the REF-RAW ambient
+				 *    ``value1``. It is the value measured by
+				 *    the ADC inside the sensor when all LEDs
+				 *    are off, so it works as a reference point
+				 *    for the first three values. However this
+				 *    time the ``value0``, ``value1`` and
+				 *    ``value2`` are sent relative to this value
+				 *    by the sensor automatically, so it is not
+				 *    necessary to use this value at all.
+				 *
 				 * @description: Raw Color Components - sets LED color to white (all LEDs rapidly cycling)
-				 * @value0: Red (0 to 1020???)
-				 * @value1: Green (0 to 1020???)
-				 * @value2: Blue (0 to 1020???)
-				 * @value3: Ambient? (0 to 1020???)
+				 * @value0: Reflected red light intensity (0 to 1023)
+				 * @value0_footnote: [#lego-ev3-color-mode4-value012]_
+				 * @value1: Reflected green light intensity (0 to 1023)
+				 * @value1_footnote: [#lego-ev3-color-mode4-value012]_
+				 * @value2: Reflected blue light intensity (0 to 1023)
+				 * @value2_footnote: [#lego-ev3-color-mode4-value012]_
+				 * @value3: ADC reading for LEDs off (reference) (0 to 1023)
```

Comment:
I've been wondering if I should revert https://github.com/ev3dev/lego-linux-drivers/commit/3f7760ec7be41813a55dfd911a529606b38e1538. The sensor itself says that this mode only has 3 values. I got the idea about the 4th value from Dexter Industries BrickPi python library. But I've been wondering it this is actually a useful value or just leftover data in a buffer. UART sensors can only send data in sizes that are a power of 2, so that could explain why the extra data is sent even if it was never meant to be used.

Have you actually tried the ev3dev-buster kernel to see if this value seems similar to the same value in the RAW-REF mode?

Reply:
I haven't tried this on ev3dev, but on stock firmware with c4ev3 it seemed to work similarly. I will recheck it. There is a simple test for this - if this value briefly changes when the intensity of external lighting illuminating the sensor changes, then this should be the ADC measurement when no LEDs are on.

I think that this value could still be useful - while one cannot measure steady ambient light, one could measure changes in ambient light this way.

Reply:
OK, if it seems useful, then let's keep it. :smile:

Reply:
I have tried it again on C4EV3 (I don't have a SD card by me... :smile:) and the fourth value indeed behaves like the ambient of REF-RAW. I think this value can be useful and it's nice that ev3dev would provide it out of the box.

Without changes in external lighting, the values move around 670. When I move my phone's flashlight in front of the sensor, it temporarily moves to zero and the returns back. When I remove the light, the value jumps up and then returns again.